Source/DafnyLanguageServer.Test/Util/TestNotificationReceiver.cs

```diff
@@ -26,46 +25,42 @@ public class TestNotificationReceiver<TNotification> {
     }
 
     public void NotificationReceived(TNotification request) {
-      logger.LogTrace($"Received {request.Stringify()}");
       Assert.NotNull(request);
-      notifications.Enqueue(request);
+      logger.LogTrace($"Received {request.Stringify()}");
+
+      // The order here is important. It's OK to have something be in the history before it's in the queue
+      // But it's not OK for something not to be in history that's already available through the queue
       notificationHistory.Add(request);
-      availableNotifications.Release();
+      notifications.Enqueue(request);
     }
 
-    public bool HasPendingNotifications => !notifications.IsEmpty;
+    public bool HasPendingNotifications => notifications.Size != 0;
 
     public IReadOnlyList<TNotification> History => notificationHistory;
 
     public void ClearHistory() {
       notificationHistory.Clear();
     }
 
-    public TNotification GetLast(Func<TNotification, bool> predicate) {
-      var result = History.LastOrDefault(predicate);
-      ClearQueue();
-      return result;
-    }
-
-    public void ClearQueue() {
-      notifications.Clear();
-      availableNotifications = new(0);
+    public TNotification? GetLast(Func<TNotification, bool> predicate) {
+      // The order here is important.
+      // It's not OK to remove items from the queue after getting the last item from history
+      while (notifications.Size > 0) {
+        _ = notifications.Dequeue(CancellationToken.None);
+      }
+      return History.LastOrDefault(predicate);
```

Comment:
I'm completely perplexed by the new logic here, but I was before by the old logic. If Notifications is a concurrent queue, how do you ensure that GetLast actually returns the last one ever, if that's the intended meaning? And what's the relationship between notifications and History?

Also, how do you prevent concurrent threads from adding something to the notification queue between the end of the while to the beginning of the return statement?

Reply:
> I'm completely perplexed by the new logic here, but I was before by the old logic.

🤣 

I think the confusion comes from `GetLast` having a difficult to explain contract. I considered inlining it, since it does not have many usages, but that would lead to a little code duplication.

GetLast does not return the last ever notification. `GetLatest` would be a better name.

History is all notifications that have been received. `TestNotificationReceiver ` implicitly remembers an index into that history, starting at index 0. Notifications is a list of notifications starting at that cursor and extending into the infinite future. You can call `notifications.Dequeue` to return the notification located at the cursor, and bump the cursor up by one.

> Also, how do you prevent concurrent threads from adding something to the notification queue between the end of the while to the beginning of the return statement?

I don't, but the entire while is a performance optimization so the clients don't iterate through 'old' notifications. If what you describe happens, then the clients may inspect and old notification, after which they'll call `AwaitNextNotificationAsync` again to get a newer one.

Reply:
Actually, I found some client code that does require the queue to be empty after calling GetLast. For example this code:

```
var diagnostics1 = diagnosticsReceiver.GetLast(documentItem);
Assert.Equal(4, diagnostics1.Length);
ApplyChange(ref documentItem, ((7, 25), (10, 17)), "");
var diagnostics2 = await GetNextDiagnostics(documentItem);
```

The `GetNextDiagnostics` call returns the wrong result if the queue was not at the position of `diagnostics1` after calling `diagnosticsReceiver.GetLast`. I've updated the code again to account for that.